File: backend/offer_engine.py
"""Promotional offer discount calculator (shared by API validation and POS)."""
from __future__ import annotations

from collections import defaultdict
from typing import Any

OFFER_TYPES = frozenset({
    "second_half",   # 1 + 50% on second (configurable %)
    "bogo",          # buy 1 get 1 free
    "buy2get1",      # buy 2 get 1 free
    "direct_percent",
    "direct_amount",
})


def _unit_lines(cart_lines: list[dict], product_ids: set[int]) -> list[dict]:
    """Flatten cart lines into per-unit entries preserving scan order."""
    units: list[dict] = []
    for line_idx, line in enumerate(cart_lines):
        pid = int(line["product_id"])
        if pid not in product_ids:
            continue
        price = float(line["unit_price"])
        qty = int(line["quantity"])
        for _ in range(qty):
            units.append({"line_idx": line_idx, "unit_price": price, "product_id": pid})
    return units


def _distribute(units: list[dict], per_unit_discounts: list[float]) -> dict[int, float]:
    by_line: dict[int, float] = defaultdict(float)
    for u, d in zip(units, per_unit_discounts):
        if d > 0:
            by_line[u["line_idx"]] += d
    return dict(by_line)


def compute_offer_line_discounts(
    offer: dict[str, Any],
    cart_lines: list[dict],
) -> tuple[dict[int, float], float]:
    """
    cart_lines: [{product_id, quantity, unit_price}, ...]
    Returns ({line_index: discount_amount}, total_savings).
    """
    product_ids = {int(p) for p in (offer.get("product_ids") or [])}
    if not product_ids:
        return {}, 0.0

    units = _unit_lines(cart_lines, product_ids)
    if not units:
        return {}, 0.0

    offer_type = offer.get("offer_type") or ""
    pct = float(offer.get("discount_percent") or 50)
    amt = float(offer.get("discount_amount") or 0)
    per_unit = [0.0] * len(units)

    if offer_type == "direct_percent":
        rate = max(0.0, min(100.0, pct)) / 100.0
        for i, u in enumerate(units):
            per_unit[i] = round(u["unit_price"] * rate, 2)

    elif offer_type == "direct_amount":
        for i, u in enumerate(units):
            per_unit[i] = round(min(amt, u["unit_price"]), 2)

    elif offer_type == "second_half":
        rate = max(0.0, min(100.0, pct)) / 100.0
        for i in range(0, len(units) - (len(units) % 2), 2):
            per_unit[i + 1] = round(units[i + 1]["unit_price"] * rate, 2)

    elif offer_type == "bogo":
        for i in range(0, len(units) - (len(units) % 2), 2):
            cheaper_idx = i if units[i]["unit_price"] <= units[i + 1]["unit_price"] else i + 1
            per_unit[cheaper_idx] = round(units[cheaper_idx]["unit_price"], 2)

    elif offer_type == "buy2get1":
        for i in range(0, len(units) - (len(units) % 3), 3):
            trio = units[i : i + 3]
            cheapest = min(range(3), key=lambda j: trio[j]["unit_price"])
            per_unit[i + cheapest] = round(trio[cheapest]["unit_price"], 2)

    else:
        return {}, 0.0

    by_line = _distribute(units, per_unit)
    total = round(sum(per_unit), 2)
    return by_line, total


def apply_offers_to_cart(
    cart_lines: list[dict],
    offers: list[dict[str, Any]],
) -> tuple[list[dict], list[int], float]:
    """
    Apply offers by priority (lower number = higher priority).
    Each product may only belong to one active offer (enforced on save).
    Returns (lines with offer_discount + offer_id), list of offer ids used, total savings.
    """
    sorted_offers = sorted(offers, key=lambda o: (int(o.get("priority") or 0), int(o.get("id") or 0)))
    line_offer_discount: dict[int, float] = defaultdict(float)
    line_offer_id: dict[int, int] = {}
    used_offer_ids: list[int] = []

    for offer in sorted_offers:
        if not offer.get("active", True):
            continue
        by_line, total = compute_offer_line_discounts(offer, cart_lines)
        if total <= 0:
            continue
        oid = int(offer["id"])
        if oid not in used_offer_ids:
            used_offer_ids.append(oid)
        for idx, disc in by_line.items():
            line_offer_discount[idx] += disc
            line_offer_id[idx] = oid

    total_savings = round(sum(line_offer_discount.values()), 2)
    enriched = []
    for idx, line in enumerate(cart_lines):
        enriched.append({
            **line,
            "offer_id": line_offer_id.get(idx),
            "offer_discount": round(line_offer_discount.get(idx, 0), 2),
        })
    return enriched, used_offer_ids, total_savings

File: frontend/src/lib/offerEngine.ts
import type { CartItem } from './api'

export type OfferType =
  | 'second_half'
  | 'bogo'
  | 'buy2get1'
  | 'direct_percent'
  | 'direct_amount'

export interface PromoOffer {
  id: number
  name_en: string
  name_ar: string
  offer_type: OfferType
  discount_percent?: number | null
  discount_amount?: number | null
  priority: number
  active: boolean
  valid_from?: string | null
  valid_to?: string | null
  branch_ids?: number[] | null
  product_ids: number[]
  products?: { id: number; name_en: string; name_ar: string; barcode: string; price: number }[]
}

export function offerTypeLabel(type: OfferType, lang: 'ar' | 'en'): string {
  const en: Record<OfferType, string> = {
    second_half: '1 + 50% on 2nd',
    bogo: 'Buy 1 Get 1 Free',
    buy2get1: 'Buy 2 Get 1 Free',
    direct_percent: 'Direct % discount',
    direct_amount: 'Direct amount off',
  }
  const ar: Record<OfferType, string> = {
    second_half: '١ + ٥٠٪ على الثاني',
    bogo: 'اشتري واحد واحصل على الثاني مجاناً',
    buy2get1: 'اشتري ٢ واحصل على الثالث مجاناً',
    direct_percent: 'خصم نسبة مباشر',
    direct_amount: 'خصم مبلغ مباشر',
  }
  return (lang === 'ar' ? ar : en)[type] || type
}

const calcManualDiscount = (
  gross: number,
  mode: 'amount' | 'percent' | undefined,
  value: number | undefined,
): number => {
  const v = Math.max(0, value || 0)
  const d = mode === 'percent' ? (gross * v) / 100 : v
  return Math.min(gross, Math.max(0, +d.toFixed(2)))
}

/** Product IDs that belong to any active promo offer group. */
export function offerProductIds(offers: PromoOffer[]): Set<number> {
  const ids = new Set<number>()
  for (const o of offers) {
    if (!o.active) continue
    for (const pid of o.product_ids || []) ids.add(pid)
  }
  return ids
}

export function isOfferProduct(productId: number, offers: PromoOffer[]): boolean {
  return offerProductIds(offers).has(productId)
}

type UnitLine = { lineIdx: number; unitPrice: number; productId: number }

function flattenUnits(lines: { product_id: number; quantity: number; unit_price: number }[], productIds: Set<number>): UnitLine[] {
  const units: UnitLine[] = []
  lines.forEach((line, lineIdx) => {
    if (!productIds.has(line.product_id)) return
    for (let q = 0; q < line.quantity; q++) {
      units.push({ lineIdx, unitPrice: line.unit_price, productId: line.product_id })
    }
  })
  return units
}

function computeOfferLineDiscounts(
  offer: PromoOffer,
  cartLines: { product_id: number; quantity: number; unit_price: number }[],
): { byLine: Map<number, number>; total: number } {
  const productIds = new Set(offer.product_ids || [])
  const units = flattenUnits(cartLines, productIds)
  if (!units.length) return { byLine: new Map(), total: 0 }

  const perUnit = new Array(units.length).fill(0)
  const pct = Number(offer.discount_percent ?? 50)
  const amt = Number(offer.discount_amount ?? 0)

  if (offer.offer_type === 'direct_percent') {
    const rate = Math.max(0, Math.min(100, pct)) / 100
    units.forEach((u, i) => { perUnit[i] = +(u.unitPrice * rate).toFixed(2) })
  } else if (offer.offer_type === 'direct_amount') {
    units.forEach((u, i) => { perUnit[i] = +Math.min(amt, u.unitPrice).toFixed(2) })
  } else if (offer.offer_type === 'second_half') {
    const rate = Math.max(0, Math.min(100, pct)) / 100
    for (let i = 0; i + 1 < units.length; i += 2) {
      perUnit[i + 1] = +(units[i + 1].unitPrice * rate).toFixed(2)
    }
  } else if (offer.offer_type === 'bogo') {
    for (let i = 0; i + 1 < units.length; i += 2) {
      const cheaper = units[i].unitPrice <= units[i + 1].unitPrice ? i : i + 1
      perUnit[cheaper] = +units[cheaper].unitPrice.toFixed(2)
    }
  } else if (offer.offer_type === 'buy2get1') {
    for (let i = 0; i + 2 < units.length; i += 3) {
      const trio = [units[i], units[i + 1], units[i + 2]]
      const cheapest = trio.reduce((best, u, j) => (u.unitPrice < trio[best].unitPrice ? j : best), 0)
      perUnit[i + cheapest] = +trio[cheapest].unitPrice.toFixed(2)
    }
  }

  const byLine = new Map<number, number>()
  units.forEach((u, i) => {
    if (perUnit[i] > 0) {
      byLine.set(u.lineIdx, (byLine.get(u.lineIdx) || 0) + perUnit[i])
    }
  })
  const total = +perUnit.reduce((s, d) => s + d, 0).toFixed(2)
  return { byLine, total }
}

export function applyOffersToCart(items: CartItem[], offers: PromoOffer[]): {
  items: CartItem[]
  offerIds: number[]
  offerSavings: number
  offerNames: string
} {
  if (!offers.length || !items.length) {
    return {
      items: items.map((i) => ({ ...i, offer_id: undefined, offer_discount: 0 })),
      offerIds: [],
      offerSavings: 0,
      offerNames: '',
    }
  }

  const lines = items.map((i) => ({
    product_id: i.product.id,
    quantity: i.quantity,
    unit_price: i.unit_price,
  }))

  const sorted = [...offers].sort((a, b) => (a.priority - b.priority) || (a.id - b.id))
  const inOfferGroup = offerProductIds(sorted)
  const lineOfferDiscount = new Map<number, number>()
  const lineOfferId = new Map<number, number>()
  const usedIds: number[] = []
  const usedNames: string[] = []

  for (const offer of sorted) {
    if (!offer.active) continue
    const { byLine, total } = computeOfferLineDiscounts(offer, lines)
    if (total <= 0) continue
    if (!usedIds.includes(offer.id)) {
      usedIds.push(offer.id)
      usedNames.push(offer.name_en)
    }
    byLine.forEach((disc, idx) => {
      lineOfferDiscount.set(idx, (lineOfferDiscount.get(idx) || 0) + disc)
      lineOfferId.set(idx, offer.id)
    })
  }

  const enriched = items.map((item, idx) => {
    const gross = item.quantity * item.unit_price
    const offerDiscount = +(lineOfferDiscount.get(idx) || 0).toFixed(2)
    const blockedManual = inOfferGroup.has(item.product.id)
    const manual = blockedManual
      ? 0
      : calcManualDiscount(gross, item.discount_mode, item.discount_value)
    const totalDiscount = Math.min(gross, +(manual + offerDiscount).toFixed(2))
    return {
      ...item,
      offer_id: lineOfferId.get(idx),
      offer_discount: offerDiscount,
      discount: totalDiscount,
    }
  })

  const offerSavings = +[...lineOfferDiscount.values()].reduce((s, d) => s + d, 0).toFixed(2)
  return { items: enriched, offerIds: usedIds, offerSavings, offerNames: usedNames.join(', ') }
}
